addons/api/addon/services/indexed-db.js

```diff
@@ -0,0 +1,40 @@
+import Service from '@ember/service';
+import Dexie from 'dexie';
+import { inject as service } from '@ember/service';
+
+// List of resources we support to be cached and searched in
+// indexedDb. Any field that can be searched upon should be added to
+// this index. Increment the version number of the database whenever
+// indexes change.
+export const modelIndexes = {
+  target: '&id, attributes.type',
+  session: '&id, attributes.type',
+};
+
+export const formatDbName = (userId, clusterUrl) =>
+  `boundary-${userId}-${clusterUrl}`;
+
+/**
+ * Service to encapsulate the IndexedDB implementation. To use this service, call
+ * `setup` from the application root and
+ */
+export default class IndexedDbService extends Service {
+  // =services
+  @service session;
+
+  // =attributes
+  #db;
+  get db() {
+    return this.#db;
+  }
+
+  setup(dbName) {
+    // Don't run setup again if we already have one or if we didn't get a name
+    if (this.#db || !dbName) {
+      return;
+    }
+
+    this.#db = new Dexie(dbName);
+    this.#db.version(1).stores(modelIndexes);
```

Comment:
curious about this version here, What do they mean? and would there be a case where we access another version?

Reply:
Take a look at the [docs](https://dexie.org/docs/Dexie/Dexie.version()) but it's similar to how actual DBs work and having to migrate when any schema changes occur.

The version is to indicate what the schema was like at a particular version and if there are any changes, use a different version. That way users can actually migrate their data to whatever the new schema is but it's a bit overkill for us.

I was hoping there was an easy way to just detect any schema changes and destroy the DB and recreate it but I haven't seen an easy way to do that yet.